tails_files/install.sh

```diff
@@ -68,23 +87,360 @@ if [ -f $XSESSION_RC ]; then
-if ! grep -q 'custom-nm-hooks' "$TAILSCFG/persistence.conf"; then
-  echo "/etc/NetworkManager/dispatcher.d	source=custom-nm-hooks,link" >> $TAILSCFG/persistence.conf
-fi
-mkdir -p $TAILSCFG/custom-nm-hooks
-cp -p $INSTALL_DIR/99-tor-reload.sh $TAILSCFG/custom-nm-hooks
-cp -p $INSTALL_DIR/99-tor-reload.sh /etc/NetworkManager/dispatcher.d/
+  # Remove previous NetworkManager hook if present. In 0.3.6 the prefix
+  # was "99-", which caused it to run too late. In 0.3.5 the prefix was "70-"
+  # and should also be removed for backwards-compatibility. There are a variety
+  # of directories the associated files were placed in, so remove from all locations,
+  # otherwise we'll have multiple redundant hooks. Under Tails v2, if you restart
+  # the tor service too rapidly in succession, it will refuse to start.
+  for d in $tails_live_persistence \
+      $securedrop_dotfiles \
+      $network_manager_dispatcher \
+      "${tails_live_persistence}/custom-nm-hooks"; do
+    for f in 70-tor-reload.sh 99-tor-reload.sh ; do
+      rm -f "${d}/${f}" > /dev/null 2>&1
+    done
+  done
+
+  # Remove binary setuid wrapper from previous tails_files installation, if it exists.
+  rm -f "${securedrop_dotfiles}/securedrop_init" 2>&1
+}
+
+
+function copy_securedrop_dotfiles()
+{
+  mkdir -p "$securedrop_dotfiles"
+
+  # copy icon, launchers and scripts
+  cp -f securedrop_icon.png $securedrop_dotfiles
+  cp -f document.desktop $securedrop_dotfiles
+  cp -f source.desktop $securedrop_dotfiles
+  cp -f securedrop_init.py $securedrop_init_script
+
+  # set permissions
+  chmod 755 $securedrop_dotfiles
+
+}
+
+function set_permissions_on_securedrop_dotfiles()
+{
+  # Follow up function to enforce correct permissions and ownership
+  # on SecureDrop dotfiles in the staging directly. Separated
+  # into function to aid readability.
+  chown root:root $securedrop_init_script
+  chmod 700 $securedrop_init_script
+  chown root:root $torrc_additions
+  chmod 400 $torrc_additions
+
+  chown amnesia:amnesia "${securedrop_dotfiles}/securedrop_icon.png"
+  chmod 600 "${securedrop_dotfiles}/securedrop_icon.png"
+}
+
+function prompt_for_document_aths_info()
+{
+  # Ad-hoc memoization via a global variable. If we've already prompted for
+  # the ATHS info here, the global will be populated. Otherwise, we'll store it.
+  if [ -z $document_aths_info_global ] ; then
+    # During initial provisioning of the Journalist Workstation, the ATHS file
+    # for the Document Interface may not be present. If not, prompt for the info.
+    aths_regex="^(HidServAuth [a-z2-7]{16}\.onion [A-Za-z0-9+/.]{22})"
+    # Loop while reading the input from a dialog box.
+    while [[ ! "$document_aths_info" =~ $aths_regex ]]; do
+      document_aths_info="$(zenity --entry \
+        --title='Hidden service authentication setup' \
+        --width=600 \
+        --window-icon=${securedrop_dotfiles}/securedrop_icon.png \
+        --text='Enter the HidServAuth value to be added to /etc/tor/torrc:')"
```

Comment:
Please make sure you are *very clear* about whether or not "the HidServAuth value" does or does not include the "HidServAuth" prefix. This is a common point of confusion for admins at this stage in the process.